### PR TITLE
Support Xgboost in NNFrame

### PR DESCRIPTION
**Pull request by dding3**

### CODE REVIEW

File: pyzoo/zoo/examples/nnframes/xgboost/XgboostExample.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#
```
Comment (shane-huang) [inline pyzoo/zoo/examples/nnframes/xgboost/XgboostExample.py:1]: We should not use this as the example. Let's use public dataset for examples. There're plenty to choose in xgboost's repo e.g. https://github.com/dmlc/xgboost/blob/master/demo/guide-python/predict_first_ntree.py

File: pyzoo/zoo/pipeline/nnframes/nn_file_reader.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#
```
Comment (shane-huang) [inline pyzoo/zoo/pipeline/nnframes/nn_file_reader.py:1]: shall we use spark.read.csv directly?

File: zoo/pom.xml [hunk 3]
```diff
@@ -303,6 +303,16 @@
...
         </dependency>
+        <dependency>
+            <groupId>ml.dmlc</groupId>
+            <artifactId>xgboost4j</artifactId>
+            <version>0.90</version>
+        </dependency>
+        <dependency>
+            <groupId>ml.dmlc</groupId>
+            <artifactId>xgboost4j-spark</artifactId>
+            <version>0.90</version>
```
Comment (shane-huang) [inline zoo/pom.xml:314]: shouldn't we use xgboost 1.0? According to yang's team 1.0 should run faster than 0.9.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/nnframes/xgboost/XgboostExample.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
+package com.intel.analytics.zoo.examples.nnframes.xgboost
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/nnframes/xgboost/XgboostExample.scala:1]: scala examples is not as important as python. I think we can do with python example for now.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala [hunk 5]
```diff
@@ -304,3 +308,46 @@ object NNClassifierModel extends MLReadable[NNClassifierModel[_]] {
     new NNClassifierModel.NNClassifierModelReader
   }
 }
+
+class NNXGBoostClassifierModel private[zoo] (
+   val model: XGBoostClassificationModel) {
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala:313]: some api docs?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNFileReader.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.zoo.pipeline.nnframes
+
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.{DataFrame, SQLContext}
+
+object NNFileReader {
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNFileReader.scala:22]: Do we need this wrapper?

File: pyzoo/zoo/examples/nnframes/xgboost/XgboostExample.py (hunk 1 above)
Comment (jason-dai) [reply pyzoo/zoo/examples/nnframes/xgboost/XgboostExample.py:1]: Don't need an example here; need to add some unit tests

File: zoo/pom.xml (hunk 3 above)
Comment (jason-dai) [reply zoo/pom.xml:314]: Yes, let's use 1.0; which Spark version does it support?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNFileReader.scala (hunk 6 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNFileReader.scala:22]: No need for this

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/nnframes/xgboost/XgboostExample.scala (hunk 4 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/nnframes/xgboost/XgboostExample.scala:1]: No need for an example here; add some unit tests

File: pyzoo/zoo/pipeline/nnframes/nn_file_reader.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+import sys
+from zoo.common.utils import callZooFunc
+
+if sys.version >= '3':
+    long = int
+    unicode = str
+
+
+class NNFileReader:
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/nnframes/nn_file_reader.py:25]: No need for this; just use spark.read.format

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala [hunk 8]
```diff
@@ -304,3 +308,46 @@ object NNClassifierModel extends MLReadable[NNClassifierModel[_]] {
     new NNClassifierModel.NNClassifierModelReader
   }
 }
+
+class NNXGBoostClassifierModel private[zoo] (
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala:312]: Change to `XGBClassifierModel`

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala [hunk 9]
```diff
@@ -304,3 +308,46 @@ object NNClassifierModel extends MLReadable[NNClassifierModel[_]] {
...
+    model.setFeaturesCol("DenseFeatures")
+    var output = model.transform(xgbInput).drop("DenseFeatures", "featureAssembledVector")
+    if(predictionCol != null) {
+      output = output.withColumnRenamed("prediction", predictionCol)
+    }
+    output
+  }
+}
+
+object NNXGBoostClassifierModel {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala:349]: `XGBClassifierModel`

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala [hunk 10]
```diff
@@ -304,3 +308,46 @@ object NNClassifierModel extends MLReadable[NNClassifierModel[_]] {
...
+    require(featuresCols!=None, "Please set feature columns before transform")
+    val featureVectorAssembler = new VectorAssembler()
+      .setInputCols(featuresCols)
+      .setOutputCol("featureAssembledVector")
+    val assembledDF = featureVectorAssembler.transform(dataset)
+
+    import org.apache.spark.sql.functions.{col, udf}
+    import org.apache.spark.ml.linalg.Vector
+    val asDense = udf((v: Vector) => v.toDense)
+    val xgbInput = assembledDF.withColumn("DenseFeatures", asDense(col("featureAssembledVector")))
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala:339]: do we need to check if featureAssembledVector is sparse?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/XGBoostWrapper.scala [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package ml.dmlc.xgboost4j.scala.spark
+
+object XGBoostWrapper {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/XGBoostWrapper.scala:3]: XGBoostHelper

File: pyzoo/zoo/pipeline/nnframes/nn_classifier.py [hunk 12]
```diff
@@ -579,3 +579,27 @@ def __init__(self, model, feature_preprocessing=None, jvalue=None,
     def load(path):
         jvalue = callZooFunc("float", "loadNNClassifierModel", path)
         return NNClassifierModel(model=None, feature_preprocessing=None, jvalue=jvalue)
+
+
+class NNXGBoostClassifierModel:
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/nnframes/nn_classifier.py:584]: `XGBClassifierModel`

File: zoo/pom.xml (hunk 3 above)
Comment (shane-huang) [reply zoo/pom.xml:314]: I think spark verison 2.4?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala (hunk 10 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNClassifier.scala:349]: Add a ut for dense assembled features and it works

**Comment by dding3** [comment]
jekins passed. http://10.239.47.210:18888/job/ZOO-PR-Validation/3335/

**Pull request merged**